application.py

```diff
@@ -1308,6 +1361,106 @@ def get(self):
         return make_payload(data)
 
 
+class EgressAPI(Resource):
+    @swagger.doc({
+        'summary': 'Download expression data for filter based on metadata filter',
+        'tags': ['download'],
+        'description': "Cells takes query parameters defined in the schema retrieved from the /initialize enpoint. "
+                       "<br>For categorical metadata keys filter based on `key=value` <br>"
+                       " For continuous metadata keys filter by `key=min,max`<br> Either value "
+                       "can be replaced by a \*. To have only a minimum value `key=min,\*`  To have only a maximum "
+                       "value `key=\*,max` <br>Graph data (if retrieved) is normalized"
+                       " To only retrieve cells that don't have a value for the key filter by `key`",
+        'responses': {
+            '200': {
+                'description': 'csv of cell by gene expression counts',
+            },
+
+            '400': {
+                'description': 'bad query params or bad download',
+            }
+        }
+    })
+    def get(self):
+        global e
+        e = ExpressionMatrix(application.config["DATA_DIR"], True)
+        try:
+            qs = parse_querystring(request.args)
+        except QueryStringError as error:
+            return make_payload({}, str(error), 400)
+        output_cellset = "outcellset_{}".format(request.query_string)
+        if not len(qs):
+            output_cellset = "AllCells"
+        try:
+            # create cellset
+            cellset_from_query(qs, output_cellset)
+            expression = e.getDenseExpressionMatrix("AllGenes", output_cellset)
+            expression = expression.transpose()
+            genes = all_genes()
+            cellids = e.getCellSet(output_cellset)
+            # transform to file
+            si = io.StringIO()
+            writer = csv.writer(si)
+            writer.writerow([""] + [get_cell_name(cid) for cid in cellids])
+            for idx, row in enumerate(expression):
```

Comment:
Recommending that you do the text conversion and output writing row-by-row (rather than converting the entire data set prior to writing, which requires a much larger StringIO buffer).   This will require much less memory when the data sets get large.